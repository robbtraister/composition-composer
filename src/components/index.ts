'use strict'

export * from './app'
export * from './root'
export * from './content'
export * from './meta'
export * from './resource'
export * from './static'
export * from './styles'
export * from './title'
export * from './tree'
