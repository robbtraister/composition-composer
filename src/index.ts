'use strict'

export * from '@composition/components'
