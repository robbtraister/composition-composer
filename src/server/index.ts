'use strict'

import app from './app'

import logger from '../utils/logger'

import { port as envPort } from '../../env'

export { app }

export function server(options: Options = {}) {
  process.on('disconnect', () => {
    logger.info(`Worker[${process.pid}] disconnected`)
  })
  process.on('exit', code => {
    logger.info(`Worker[${process.pid}] exited with code: ${code}`)
  })

  const port = Number(options.port) || envPort

  return app(options).listen(port, err =>
    err
      ? console.error(err)
      : logger.info(`Worker[${process.pid}] listening on port: ${port}`)
  )
}

export function main(options: Options = {}) {
  process.title = `Composition Server [${process.pid}]`
  return server(options)
}

export default server

// use eval and __filename instead of module to preserve functionality in webpack artifact
// eslint-disable-next-line no-eval
const isScript = eval('require.main && (require.main.filename === __filename)')
isScript && main()
