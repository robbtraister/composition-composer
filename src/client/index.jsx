'use strict'

/* global Composition */

import React from 'react'
import ReactDOM from 'react-dom'

import { Page } from '../components'

import unpack from '../utils/unpack'

function getComponent(type) {
  return unpack(Composition.components[type]) || type
}

function getContent({ source, query }) {
  const contentPromise = window
    .fetch(
      `/api/content/fetch?source=${encodeURIComponent(
        source
      )}&query=${encodeURIComponent(JSON.stringify(query))}`
    )
    .then(res => {
      const expires = res.headers.get('expires')
      contentPromise.expires = expires && +new Date(expires)
      return res.json()
    })
    .then(value => {
      contentPromise.value = value
      return value
    })
  return contentPromise
}

const resolutions = {}
async function resolve(uri) {
  if (!(uri in resolutions)) {
    resolutions[uri] = window
      .fetch(
        `/api/resolve?uri=${encodeURIComponent(
          uri
        )}&output=${encodeURIComponent(Composition.output)}`
      )
      .then(res => res.json())
  }
  return resolutions[uri]
}

function render() {
  if (!Composition.tree) return

  const renderElement = window.document.getElementById('composition-app')
  if (!renderElement) return

  const serverHtml = renderElement.innerHTML

  const styleElement = window.document.getElementById('composition-app-styles')
  const serverStyleHref = styleElement && styleElement.href
  if (styleElement) {
    styleElement.href = `/dist/styles/templates/${Composition.styleHash}.css`
  }

  try {
    ReactDOM[Composition.method || 'render'](
      <Page
        getComponent={getComponent}
        getContent={getContent}
        cache={Composition.cache}
        location={
          window.location.pathname +
          window.location.search +
          window.location.hash
        }
        output={Composition.output}
        resolve={resolve}
        single-page={Composition.singlePage}
        template={Composition.template}
        tree={Composition.tree}
      />,
      renderElement
    )
  } catch (e) {
    console.error(e)
    if (styleElement) {
      styleElement.href = serverStyleHref
    }
    renderElement.innerHTML = serverHtml
  }
}

window.document.addEventListener('DOMContentLoaded', render)
