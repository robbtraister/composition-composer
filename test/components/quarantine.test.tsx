'use strict'

/* global expect, test */

import assert from 'assert'

import React from 'react'
import { render } from '@testing-library/react'

import { Root, useComponentContext, useRootContext } from '../../src'

function FailFunction() {
  // random logic to waste time
  let x = 0
  for (let i = 0; i < 10000000; i++) {
    x += i ** 2
  }

  // can't throw by default, so make math that is guaranteed to be truthy
  if (0 * x === 0) {
    throw new Error('component render failed')
  }
  return <React.Fragment />
}

class FailClass extends React.Component {
  render() {
    return (
      <>
        text
        <FailFunction />
      </>
    )
  }
}

function Success({ children }: { children: React.ElementType }) {
  const { id, type } = useComponentContext()
  const { location } = useRootContext()

  assert.equal(type, 'div')
  assert.equal(location, '/test')

  return (
    <div data-type={type} data-id={id}>
      {children}
    </div>
  )
}

function getComponent(type: string): React.ComponentType {
  return (
    {
      'fail-class': FailClass,
      'fail-function': FailFunction
    }[type] || Success
  )
}

function getFailTree(type) {
  return {
    type: 'div',
    id: 'abc',
    children: [
      {
        type,
        id: 'def',
        children: [
          {
            type: 'div',
            id: 'ghi'
          }
        ]
      },
      {
        type: 'div',
        id: 'tuv',
        children: [
          {
            type: 'div',
            id: 'xyz'
          }
        ]
      }
    ]
  }
}

test('Quarantine Function Component', () => {
  const { asFragment } = render(
    <Root
      getComponent={getComponent}
      location="/test"
      tree={getFailTree('fail-function')}
      quarantine
    />
  )
  expect(asFragment()).toMatchSnapshot()
})

test('Quarantine Class Component', () => {
  const { asFragment } = render(
    <Root
      getComponent={getComponent}
      location="/test"
      tree={getFailTree('fail-class')}
      quarantine
    />
  )
  expect(asFragment()).toMatchSnapshot()
})
