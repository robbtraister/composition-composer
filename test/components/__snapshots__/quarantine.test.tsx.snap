// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Quarantine Class Component 1`] = `
<DocumentFragment>
  <div
    data-id="abc"
    data-type="div"
  >
    <div
      data-error-component="fail-class"
      data-error-id="def"
      data-error-message="component render failed"
      style="display: none;"
    />
    <div
      data-id="tuv"
      data-type="div"
    >
      <div
        data-id="xyz"
        data-type="div"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`Quarantine Function Component 1`] = `
<DocumentFragment>
  <div
    data-id="abc"
    data-type="div"
  >
    <div
      data-error-component="fail-function"
      data-error-id="def"
      data-error-message="component render failed"
      style="display: none;"
    />
    <div
      data-id="tuv"
      data-type="div"
    >
      <div
        data-id="xyz"
        data-type="div"
      />
    </div>
  </div>
</DocumentFragment>
`;
